src/boot/pe.c

```diff
@@ -441,17 +469,36 @@ EFI_STATUS pe_kernel_info(const void *base, uint32_t *ret_compat_address, size_t
         if (!verify_pe(dos, pe, /* allow_compatibility= */ true))
                 return EFI_LOAD_ERROR;
 
+        uint64_t image_base;
+        if (pe->OptionalHeader.Magic == OPTHDR32_MAGIC)
+                image_base = pe->OptionalHeader.ImageBase32;
+        else
+                image_base = pe->OptionalHeader.ImageBase64;
+
+        const PeImageDataDirectory *data_directory;
+        if (pe->OptionalHeader.Magic == OPTHDR32_MAGIC)
+                data_directory = pe->OptionalHeader.DataDirectory32;
+        else
+                data_directory = pe->OptionalHeader.DataDirectory64;
+
+        if (data_directory[5].Size != 0)
+                return log_error_status(EFI_LOAD_ERROR, "Inner kernel image contains base relocations, which we do not support.");
+
         /* When allocating we need to also consider the virtual/uninitialized data sections, so parse it out
          * of the SizeOfImage field in the PE header and return it */
-        if (ret_size_in_memory)
-                *ret_size_in_memory = pe->OptionalHeader.SizeOfImage;
+        size_t size_in_memory = pe->OptionalHeader.SizeOfImage;
 
         /* Support for LINUX_INITRD_MEDIA_GUID was added in kernel stub 1.0. */
         if (pe->OptionalHeader.MajorImageVersion < 1)
                 return EFI_UNSUPPORTED;
 
         if (pe->FileHeader.Machine == TARGET_MACHINE_TYPE) {
-                *ret_compat_address = 0;
+                if (ret_size_in_memory)
+                        *ret_size_in_memory = size_in_memory;
+                if (ret_address)
+                        *ret_address = pe->OptionalHeader.AddressOfEntryPoint;
+                if (ret_image_base)
+                        *ret_image_base = image_base;
```

Comment:
you need to initialize ret_compat_address here to zero too, if non-NULL

Reply:
doesn't really matter, but maybe fill in the params here in the same order as they appear in the function prototype